docs/website/contents/for-developers/HardForkCombinatorRollbackMonotonicityProposal.md

```diff
@@ -0,0 +1,285 @@
+
+# Introduction
+
+The Babbage->Conway transition on an earlier iteration of SanchoNet revealed a bug in the Consensus layer by which the Hard Fork Combinator transitioned to Conway because the relevant Babbage protocol parameter update proposal received sufficient votes but then that same proposal wasn't actually enacted: the major protocol version remained unchanged even after Conway started.
+
+Troubleshooting this bug and moreover determining how to fix it in such a way that makes similar mistakes less likely in the future has lead to the following mini-framework for understanding the HFC's behavior on a particular chain.
+
+So many dimensions arose from investiating such a relatively simple misbehavior for two reasons.
+First, the mechanisms implementing these policies are---at least in the current implementation---somewhat interdependent.
+Second, the HFC (including its instantiation for Cardano) remains underdocumented since back when it was originaly developed under time-pressure, and our team no longer includes its inventor.
+The Consensus Team learns new things and unfortunately re-learns some old things every time they dive in to alter it.
+(Optimistically, Nicolas Frisby feels they currently understand it now better than they have ever before.)
+
+## Dimensions
+
+The following discrete variables characterize dimensions of the HFC design space.
+In this context, the _determination_ is the `singleEraTransition` method: it's how an era's ledger state determines when the era ends.
+
+- DST - Determination is Sensitive to Ticks  - whether ticking the ledger state can change its determination of the end of its era: Sensitive vs Insensitive
+
+- CCT - Criss-Crossed Ticks - whether an era's tick rules might be used to tick across slots that are in _other_ eras: Allowed vs Disallowed
+
+- PIL - Potential to Ignore the Ledger - whether the HFC can ignore the ledger's hard fork-related governance: Possible vs Impossible
+
+- DSR - Determination is Sensitive to Rollback - for chains C and D where C < D and len(C) <= len(C /\ D) + k, whether the current era of C can have a determined end on C but an undetermined or different end on D: Sensitive vs Insensitive
+
+- ZEE - Zero Epoch Eras - whether an era (in its entirety) can contain zero epochs: Allowed vs Disallowed
+
+- ZBE - Zero Block Eras - whether an era (in its entirety) can contain zero blocks: Allowed vs Disallowed
+
+- QSR - Queries are Sensitivity to Rollback - same as DSR but for HFC-related Node-To-Client queries instead of for validity of blocks: Sensitive vs Insensitive
+
+Note the following relations.
+
+- ZEE=Allowed implies ZBE=Allowed.
+- DSR=Insensitive implies QSR=Insensitive.
+
+## Table for the Current Cardano HFC
+
+The following table and associated notes classifies the current HFC implementation in terms of these dimensions.
+
+| DST | CCT | PIL | DSR | ZEE | ZBE | QSR |
+| --- | --- | --- | --- | --- | --- | --- |
+| I   | A   | P 1 | I   | 2   | 3   | I   |
```

Comment:
About `DST: I`: I would have expected that it is sensitive to ticks since ticking might cause votes to be counted and a hardfork could be decided as a consequence of this. Would it be possible to explain why this is not the case?

Reply:
Yeah. Extremely high-level summary: it's because of PIL=Possible. The ticking does make the votes happening, but if there hasn't been enough blocks, the Cardano HFC logic will "veto" the decision and simply not transition to the next era.

Reply:
> The ticking itself does make the vote-counting actually happen

Oh, I thought this was the case. Do you know how we currently determine if we should transition to the next era based on the ledger info? I guess that we do not rely on this rule on epoch change. Do we roll out our own logic (that totally ignores the ledger?).

![image](https://github.com/input-output-hk/ouroboros-consensus/assets/175315/5768fda4-6392-40e7-8d25-ba51ed987bfe)

Reply:
> > The ticking itself does make the vote-counting actually happen
> 
> Oh, I thought this was the case.

I sounds like maybe you read "doesn't" because I wrote "does", which is a bit unusual.

> Do we roll out our own logic (that totally ignores the ledger?)

Here's the entrypoint to the Consensus-side logic https://github.com/input-output-hk/ouroboros-consensus/blob/51da3876c01edc2eec250fdc998f6cb33cdc4367/ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/ShelleyHFC.hs#L130-L164

There's probably plenty of redundancy that could be removed from this Consensus-side logic (in the "`Inspect`" module), but it's only exactly this one line that "ignores the ledger" https://github.com/input-output-hk/ouroboros-consensus/blob/51da3876c01edc2eec250fdc998f6cb33cdc4367/ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/ShelleyHFC.hs#L160 `shelleyAfterVoting` is from the ledger state; it's incremented ([here](https://github.com/input-output-hk/ouroboros-consensus/blob/51da3876c01edc2eec250fdc998f6cb33cdc4367/ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs#L428-L432)) whenever we apply a block.

Reply:
> I sounds like maybe you read "doesn't" because I wrote "does", which is a bit unusual.

I can't read :facepalm: Sorry about that.

Reply:
Thanks for the links to the relevant snippets :pray: 

> but it's only exactly this one line that "ignores the ledger"

"Ignores the ledger when there aren't `k` blocks after a transition to an era was voted on the ledger, right?